File: app/src/main/java/com/sumian/sleepdoctor/account/model/AccountViewModel.java
package com.sumian.sleepdoctor.account.model;

import android.app.Application;
import android.arch.lifecycle.AndroidViewModel;
import android.arch.lifecycle.LiveData;
import android.arch.lifecycle.MutableLiveData;
import android.support.annotation.MainThread;
import android.support.annotation.NonNull;
import android.text.TextUtils;

import com.blankj.utilcode.util.SPUtils;
import com.sumian.sleepdoctor.account.bean.Token;
import com.sumian.sleepdoctor.account.bean.UserInfo;
import com.sumian.sleepdoctor.doctor.bean.Doctor;
import com.sumian.sleepdoctor.utils.JsonUtil;

/**
 * Created by jzz
 * on 2018/1/16.
 * desc:
 */

public class AccountViewModel extends AndroidViewModel {

    private static final String SP_KEY_TOKEN = "token";
    private MutableLiveData<Token> mTokenLiveData = new MutableLiveData<>();
    private MutableLiveData<Boolean> mTokenIsInvalid = new MutableLiveData<>();

    public AccountViewModel(@NonNull Application application) {
        super(application);
    }

    public void loadTokenFromSp() {
        String tokenJson = SPUtils.getInstance().getString(SP_KEY_TOKEN, null);
        Token token = JsonUtil.fromJson(tokenJson, Token.class);
        mTokenLiveData.setValue(token);
    }

    public LiveData<Token> getLiveDataToken() {
        return mTokenLiveData;
    }

    public Token getToken() {
        return mTokenLiveData.getValue();
    }

    public UserInfo getUserInfo() {
        Token token = getToken();
        if (token == null) {
            return null;
        }
        return token.user;
    }

    public void updateBoundDoctor(Doctor doctor) {
        Token token = getToken();
        token.is_new = false;
        UserInfo userProfile = getUserInfo();
        if (doctor != null) {
            userProfile.doctor_id = doctor.getId();
        }
        userProfile.doctor = doctor;
        token.user = userProfile;
        updateUserInfo(userProfile);
    }

    public void updateToken(Token token) {
        mTokenLiveData.setValue(token);
        updateTokenInvalidState(token == null);
        persistentTokenInSp();
    }

    @MainThread
    public void updateUserInfo(UserInfo userInfo) {
        Token token = getToken();
        token.is_new = false;
        token.user = userInfo;
        updateToken(token);
    }

    public String getTokenString() {
        return getToken() == null ? null : getToken().token;
    }

    public void updateTokenInvalidState(boolean tokenIsInvalid) {
        mTokenIsInvalid.setValue(tokenIsInvalid);
    }

    public LiveData<Boolean> getLiveDataTokenInvalidState() {
        return mTokenIsInvalid;
    }

    private void persistentTokenInSp() {
        SPUtils.getInstance().put(SP_KEY_TOKEN, JsonUtil.toJson(mTokenLiveData.getValue()));
    }

    public String getLeanCloudId() {
        Token token = getToken();
        if (token == null) {
            return null;
        }
        UserInfo user = token.user;
        if (user == null) {
            return null;
        }
        return user.leancloud_id;
    }

    public boolean isLogin() {
        Token token = getToken();
        return token != null && !TextUtils.isEmpty(token.getToken());// && token.getExpired_at() * 1000L > System.currentTimeMillis();
    }

    public boolean isHaveUserInfoAndSleepBarrierTest() {
        UserInfo userInfo = getUserInfo();
        return userInfo != null && userInfo.isHaveUserInfoAndSleepBarrierTest();
    }

    public boolean isHaveAnswers() {
        UserInfo userInfo = getUserInfo();
        return userInfo != null && userInfo.isHaveAnswers();
    }

    public boolean isHaveFullUserInfo() {
        UserInfo userInfo = getUserInfo();
        return userInfo != null && userInfo.isHaveFullUserInfo();
    }
}

File: doctor/src/main/java/com/sumian/sddoctor/account/AccountViewModel.kt
package com.sumian.sddoctor.account

import android.app.Application
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import com.blankj.utilcode.util.SPUtils
import com.sumian.sddoctor.login.login.bean.DoctorInfo
import com.sumian.sddoctor.login.login.bean.LoginResponse
import com.sumian.sddoctor.util.JsonUtil

@Suppress("UNUSED_PARAMETER")
/**
 * <pre>
 *     @author : Zhan Xuzhao
 *     e-mail : xuzhao.z@sumian.com
 *     time   : 2018/6/14 10:42
 *     desc   :
 *     version: 1.0
 * </pre>
 */
class AccountViewModel(application: Application) : AndroidViewModel(application) {

    private var mTokenInfo: MutableLiveData<TokenInfo> = MutableLiveData()
    private var mDoctorInfo: MutableLiveData<DoctorInfo> = MutableLiveData()
    private var mTokenIsInvalid: MutableLiveData<Boolean> = MutableLiveData()

    init {
        val tokenInfoJson = SPUtils.getInstance().getString(KEY_TOKEN_INFO)
        val tokenInfo = JsonUtil.fromJson(tokenInfoJson, TokenInfo::class.java)
        updateTokenInfo(tokenInfo, false)
        val doctorInfoJson = SPUtils.getInstance().getString(KEY_DOCTOR_INFO)
        val doctorInfo = JsonUtil.fromJson(doctorInfoJson, DoctorInfo::class.java)
        updateDoctorInfo(doctorInfo, false)
    }

    companion object {
        private const val KEY_TOKEN_INFO: String = "token_info"
        private const val KEY_DOCTOR_INFO: String = "doctor_info"
    }

    fun getToken(): String? {
        return mTokenInfo.value?.token
    }

    fun updateTokenInfo(tokenInfo: TokenInfo?, persist: Boolean = true) {
        mTokenInfo.value = tokenInfo
        if (persist) {
            SPUtils.getInstance().put(KEY_TOKEN_INFO, JsonUtil.toJson(tokenInfo))
        }
    }

    fun updateDoctorInfo(doctorInfo: DoctorInfo?, persist: Boolean = true) {
        mDoctorInfo.value = doctorInfo
        if (persist) {
            SPUtils.getInstance().put(KEY_DOCTOR_INFO, JsonUtil.toJson(doctorInfo))
        }
    }

    fun updateTokenInfoAndDoctorInfo(response: LoginResponse?) {
        updateTokenInfo(response?.getTokenInfo())
        updateDoctorInfo(response?.doctor)
    }

    fun getTokenInfo(): LiveData<TokenInfo> {
        return mTokenInfo
    }

    fun getDoctorInfo(): LiveData<DoctorInfo> {
        return mDoctorInfo
    }

    fun updateTokenInvalidState(tokenIsInvalid: Boolean) {
        mTokenIsInvalid.postValue(tokenIsInvalid)
    }

    fun getLiveDataTokenInvalidState(): LiveData<Boolean> {
        return mTokenIsInvalid
    }

    fun logout() {
        updateTokenInfo(null, true)
    }

    fun isAuthenticated(): Boolean {
        return getAuthenticateStatus() == 2
    }

    fun getAuthenticateStatus(): Int {
        return mDoctorInfo.value?.reviewStatus ?: 0
    }

    fun isVisitorAccount(): Boolean {
        return mDoctorInfo.value?.isVisitorAccount() == true
    }

    fun isLogin(): Boolean {
        return getToken() != null
    }
}